### PR TITLE
Improved FittingTest::ChiSquared

### PR DESCRIPTION
**Pull request by regislebrun**
Now, it is based on a pure C++ implementation, with a cleaver construction of bins than the one proposed in the rot R package. It also improves a lot the execution time of the test.

### CODE REVIEW

**Comment by jschueller** [comment]
needs a rebase, the diff is unreadable because of the reindent

**Comment by regislebrun** [comment]
@jschueller done

File: lib/src/Uncertainty/StatTests/FittingTest.cxx [hunk 1]
```diff
@@ -375,85 +375,83 @@ TestResult FittingTest::ChiSquared(const Sample & sample,
...
-  if (size < 2 * nMin) throw InvalidArgumentException(HERE) << "Error: ChiSquared test cannot be used with a sample size smaller than " << 2 * nMin << ". Reduce the value of \"FittingTest-ChiSquaredMinFrequency\" below " << size / 2 << " if you really want to do this test.";
-  // The test statistics
-  Scalar xi = 0.0;
+  if (size < 2 * nMin) throw InvalidArgumentException(HERE) << "Error: ChiSquared test cannot be used with a sample size smaller than " << 2 * nMin << ". Reduce the value of \"FittingTest-ChiSquaredMinimumBinCount\" below " << size / 2 << " if you really want to use this test.";
   // Sort the data
   const Sample sortedSample(sample.sort(0));
-  const UnsignedInteger iMax = static_cast<UnsignedInteger>(round(size / nMin));
-  UnsignedInteger dataIndex = 0;
-  for (UnsignedInteger i = 1; i <= iMax; ++i)
+  UnsignedInteger binNumber;
```
Comment (jschueller) [inline lib/src/Uncertainty/StatTests/FittingTest.cxx]: I dont like uninitialized vars, even if it's set in the if branch, set it to 0

File: lib/src/Uncertainty/StatTests/FittingTest.cxx [hunk 2]
```diff
@@ -375,85 +375,83 @@ TestResult FittingTest::ChiSquared(const Sample & sample,
...
+  // If the bin number was too high to get only nMin points per bin on average, reduce the number of bins to get only nonempty bins
+  Point ticks(0);
+  Point binProbabilities(0);
+  if (distribution.isContinuous())
+  {
+    binNumber = size / nMin;
+    if (binNumber <= estimatedParameters + 1) throw InvalidArgumentException(HERE) << "Error: the bin number=" << binNumber << " must be at least equal to the number of estimated parameters + 2=" << estimatedParameters + 2 << " to apply ChiSquared test.";
```
Comment (jschueller) [inline lib/src/Uncertainty/StatTests/FittingTest.cxx]: I'd have written `binNumber < estimatedParameters + 2`

File: lib/src/Uncertainty/StatTests/FittingTest.cxx [hunk 3]
```diff
@@ -375,85 +375,83 @@ TestResult FittingTest::ChiSquared(const Sample & sample,
...
-  resultFile >> pValue;
-
-  // Clean-up everything
-  Os::Remove(dataFileName);
-  Os::Remove(resultFileName);
-  Os::Remove(commandFileName);
-
-  TestResult result(testType, testResult, pValue, pThreshold);
+    if (binNumber <= 1) throw InvalidArgumentException(HERE) << "Error: the adjusted bin number=" << binNumber << " must be at least equal to 2.";
+    if (binNumber <= estimatedParameters + 1) throw InvalidArgumentException(HERE) << "Error: the bin number=" << binNumber << " must be at least equal to the number of estimated parameters + 2=" << estimatedParameters + 2 << " to apply ChiSquared test.";
```
Comment (jschueller) [inline lib/src/Uncertainty/StatTests/FittingTest.cxx]: here too

Comment (regislebrun) [reply lib/src/Uncertainty/StatTests/FittingTest.cxx]: @jschueller done

**Comment by jschueller** [comment]
some doctest fail

**Comment by regislebrun** [comment]
@jschueller done

File: lib/src/Uncertainty/StatTests/FittingTest.cxx [hunk 4]
```diff
@@ -375,85 +375,83 @@ TestResult FittingTest::ChiSquared(const Sample & sample,
...
+      frequency += w;
+      ++index;
+    }
+    testStatistics += std::pow(frequency - probability, 2.0) / probability;
+    LOGDEBUG(OSS() << "Bin number=" << i << ", bound=" << bound << ", probability=" << probability << ", frequency=" << frequency << ", test statistic=" << testStatistics);
+  } // i
+  testStatistics *= size;
+  // Use the asymptotic statistics corrected from the number of estimated parameters
+  const Scalar pValue = DistFunc::pGamma(0.5 * (binNumber - (estimatedParameters + 1)), 0.5 * testStatistics, true);
+  TestResult result(OSS(false) << "ChiSquare" << distribution.getClassName(), (pValue > level), pValue, level);
```
Comment (jschueller) [inline lib/src/Uncertainty/StatTests/FittingTest.cxx]: distribution.getClassName() retourne tout le temps "Distribution" ici

**Comment by regislebrun** [comment]
I added a reference test based on the R documentation of chisq.tst, and an optional check of the compatibility of the given sample and the distribution support in the discrete case.

**Comment by sofianehaddad** [comment]
you merged it too quickly, I was reviewing...

**Comment by jschueller** [comment]
sorry, I was taught you were sick :)

**Comment by sofianehaddad** [comment]
that s true :-/ but I already started the review on Friday

**Pull request merged**